src/mlpack/core/data/load_image_impl.hpp

```diff
@@ -0,0 +1,305 @@
+             STBI_grey);
+  }
+  else
+  {
+    image = stbi_load(fileName.c_str(),
+             &tempWidth,
+             &tempHeight,
+             &tempChannels,
+             STBI_rgb);
+  }
+
+  if (tempWidth <= 0 || tempHeight <= 0)
+  {
+    std::ostringstream oss;
+    oss << "Image '" << fileName << "' not found." << std::endl;
+    free(image);
+    throw std::runtime_error(oss.str());
+    return false;
+  }
+  width = tempWidth;
+  height = tempHeight;
+  channels = tempChannels;
+  size_t size = width * height * channels;
+
+  // Copy image into armadillo Mat.
+  outputMatrix = arma::Mat<unsigned char>(image, size, 1, true, true);
+
+  // Free the image pointer.
+  free(image);
+  return true;
+}
+
+bool Image::Load(const std::string& fileName,
+                 bool flipVertical,
+                 arma::Mat<unsigned char>&& outputMatrix)
+{
+  size_t width, height;
+  bool status = Load(fileName, flipVertical, std::move(outputMatrix),
+      width, height, channels);
+  if (!status)
+    return status;
+
+  Log::Info << "Image width: " << width <<"  height: "<< height <<" channels: "
+      << channels << std::endl;
+
+  // Throw error if the image is incompatible with the matrix.
+  if (maxWidth > 0 && maxHeight > 0 &&
+     (width != maxWidth || height != maxHeight))
+  {
+    std::ostringstream oss;
+    oss << "Image '" << fileName;
+    oss << "' does not match matrix height or width." << std::endl;
+    oss << "Image width: " << width;
+    oss << ", Image height: " << height << std::endl;
+    oss << "Max Width: " << maxWidth;
+    oss << ", Max Height: " << maxHeight << std::endl;
+    throw std::runtime_error(oss.str());
+  }
+  return status;
+}
+
+bool Image::Load(const std::vector<std::string>& files,
+                 bool flipVertical,
+                 arma::Mat<unsigned char>&& outputMatrix)
+{
+  if (files.size() == 0)
+  {
+    std::ostringstream oss;
+    oss << "File vector is empty." << std::endl;
+    throw std::runtime_error(oss.str());
+  }
+
+  arma::Mat<unsigned char> img;
+  size_t width, height;
+  bool status = Load(files[0], flipVertical, std::move(img),
+                     width, height, channels);
+  Log::Info << "Loaded " << files[0] << std::endl;
+
+  // Decide matrix dimension using the image height and width.
+  maxWidth = std::max(maxWidth, width);
+  maxHeight = std::max(maxHeight, height);
+
+  outputMatrix.set_size(maxWidth*maxHeight*channels, files.size());
+  outputMatrix.col(0) = img;
+
+  for (size_t i = 1; i < files.size() ; i++)
+  {
+    status &= Load(files[i],
+                   flipVertical,
+                   std::move(outputMatrix.col(i)));
+    Log::Info << "Loaded " << files[i] << std::endl;
+  }
+  return status;
+}
+
+
+bool Image::Save(const std::string& fileName,
+                 size_t width,
+                 size_t height,
+                 size_t channels,
+                 bool flipVertical,
+                 arma::Mat<unsigned char>&& inputMatrix)
```

Comment:
Couldn't we  derive the number of channels from ``width, height`` and ``n_elem``?

Reply:
Yes we can but image shape usually means width, height, channels. However if you want I will change.

Reply:
Ideally we could derive width, height, channels, from the datatype, which I think is possible if we could use ``arma::cube, or arma::filed``, since most people just want to load/save the object the have. But I think for arma::mat this is difficult. Why not just provide two interface one that allows setting the channels and another one which derives the number of channels from the rest. Does that sound reasonable to you?

Reply:
Should 
I also add support for cube and field?

Reply:
I think for now ``arma::mat`` is fine.